lib/internal/warn.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+// A process Warning is similar in structure to an Error except that they
+// are emitted by the process.on('warn') event rather than being thrown.
+
+const _message = Symbol('message');
+const _name = Symbol('name');
+const prefix = `(${process.release.name}:${process.pid}) `;
+
+class Warning {
+  constructor(name, message, constructor) {
+    this[_name] = name;
+    this[_message] = `${prefix}${message}`;
+    Error.captureStackTrace(this, constructor || Warning);
+  }
+
+  get message() {
+    return this[_message];
+  }
+
+  get name() {
+    return this[_name];
+  }
+}
+
+function emit(warning) {
+  process.nextTick(() => {
+    process.emit('warn', warning);
+  });
```

Comment:
I'm not sure if emitting it asynchronously is actually a good idea; that arguably detaches it too much from the call site.

(On the other hand, synchronous events can result in unbounded recursion if the event handler itself generates warnings.)